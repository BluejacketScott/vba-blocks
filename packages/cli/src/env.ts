import { homedir } from 'os';
import unixJoin from './utils/unix-join';
import { Reporter, reporter } from './reporter';

export interface Env {
  isWindows: boolean;
  cwd: string;
  values: any;

  addins: string;
  scripts: string;
  cache: string;
  registry: string;
  packages: string;
  sources: string;
  staging: string | null;

  reporter: Reporter;
}

const cache = unixJoin(homedir(), '.vba-blocks');
const env: Env = {
  isWindows: process.platform === 'win32',
  cwd: process.cwd(),
  values: process.env,

  addins: unixJoin(__dirname, '../../addin/build'),
  scripts: unixJoin(__dirname, '../scripts'),
  cache,
  registry: unixJoin(cache, 'registry'),
  packages: unixJoin(cache, 'packages'),
  sources: unixJoin(cache, 'sources'),
  staging: null,

  reporter
};

export default env;
