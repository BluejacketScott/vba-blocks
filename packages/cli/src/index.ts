export { default as env } from './env';
export {
  Version,
  Dependency,
  Feature,
  Reference,
  Source,
  Target,
  Manifest,
  loadManifest
} from './manifest';
export { Config, loadConfig } from './config';
export { default as resolve } from './resolve';
