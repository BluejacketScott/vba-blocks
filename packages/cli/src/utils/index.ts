export { default as checksum } from './checksum';
export { default as convertToToml } from './convert-to-toml';
export { default as copy } from './copy';
export { default as download } from './download';
export * from './git';
export { default as has } from './has';
export * from './is';
export { default as last } from './last';
export { default as optionList } from './option-list';
export { default as parallel } from './parallel';
export { default as plural } from './plural';
export { default as run, escape } from './run';
export { default as tmpFile } from './tmp-file';
export { default as unique } from './unique';
export { default as walk } from './walk';
export * from './zip';
