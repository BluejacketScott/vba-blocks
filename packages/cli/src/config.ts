import { join } from 'path';
import { parse as parseToml } from 'toml';
import env from './env';
import { pathExists, readFile } from './utils';
import {
  Sources,
  RegistrySource,
  PathSource,
  GitSource,
  UnsupportedSource
} from './sources';

export type Registry =
  | {}
  | { [name: string]: { index: string; packages: string } };

export interface Flags {
  git?: boolean;
  path?: boolean;
}

export interface Config {
  registry: Registry;
  flags: Flags;
  sources: Sources;
}

export interface ConfigValue {
  registry?: Registry;
  flags?: Flags;
}

const empty: ConfigValue = { registry: {}, flags: {} };
const defaults: ConfigValue = {
  registry: {
    'vba-blocks': {
      index: 'https://github.com/vba-blocks/registry',
      packages: 'https://packages.vba-blocks.com'
    }
  },
  flags: { git: true, path: true }
};

/**
 * Load config, from local, user, and environment values
 *
 * - Search for .vba-blocks/config.toml up from cwd
 * - Load ~/.vba-blocks/config.toml
 * - Load VBA_BLOCKS_* from environment
 */
export async function loadConfig(): Promise<Config> {
  const user: ConfigValue = {
    ...empty,
    ...((await readConfig(env.cache)) || {})
  };
  const file = await findConfig(env.cwd);
  const local: ConfigValue = {
    ...empty,
    ...file ? await readConfig(file) : {}
  };
  const override = loadConfigFromEnv();

  const registry: Registry = {
    ...defaults.registry,
    ...user.registry,
    ...local.registry,
    ...override.registry
  };
  const flags: Flags = {
    ...defaults.flags,
    ...user.flags,
    ...local.flags,
    ...override.flags
  };

  const sources: Sources = {
    registry: {},
    git: flags.git ? new GitSource() : new UnsupportedSource('git'),
    path: flags.path ? new PathSource() : new UnsupportedSource('path')
  };

  for (const [name, { index, packages }] of Object.entries(registry)) {
    sources.registry[name] = new RegistrySource({
      name,
      index,
      packages
    });
  }

  return { registry, flags, sources };
}

// Read config from dir (if present)
export async function readConfig(
  dir: string
): Promise<ConfigValue | undefined> {
  const file = join(dir, 'config.toml');
  if (!await pathExists(file)) return {};

  const raw = await readFile(file);
  const parsed = parseToml(raw.toString());

  return parsed;
}

// Find config up from and including given dir
// (looking for .vba-blocks/config.toml)
export async function findConfig(dir: string): Promise<string | undefined> {
  // TODO Search from .vba-blocks/config.toml starting at cwd
  return;
}

// Load config values from environment
export function loadConfigFromEnv(): ConfigValue {
  // TODO Load override config from VBA_BLOCKS_* env variables
  return {};
}
