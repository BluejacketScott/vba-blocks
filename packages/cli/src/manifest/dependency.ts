import { Version } from './version';
import { isString } from '../utils';

export interface Dependency {
  name: string;
  defaultFeatures: boolean;
  features: string[];
  optional?: boolean;

  version?: Version;
  path?: string;
  git?: string;
  tag?: string;
  branch?: string;
  rev?: string;
}

interface RawDependency {
  'default-features': boolean;
  features: string[];
  optional?: boolean;

  version?: Version;
  path?: string;
  git?: string;
  tag?: string;
  branch?: string;
  rev?: string;
}

const EXAMPLE = `Example vba-block.toml:

  [dependencies]
  a = "^1.0.0"
  b = { version = "^0.1.0", optional = true }
  c = { path = "packages/c" }
  d = { git = "https://github.com/author/d" }
  e = { git = "https://github.com/author/e", branch = "next" }
  f = { git = "https://github.com/author/f", tag = "v1.0.0" }
  g = { git = "https://github.com/author/g", rev = "a1b2c3d4" }

  [dependencies.h]
  version = "^2.0.0"
  default-features = false
  features ["a", "b"]`;

export function parseDependencies(value: any): Dependency[] {
  return Object.entries(value).map(([name, value]) =>
    parseDependency(name, value)
  );
}

export function parseDependency(
  name: string,
  value: Version | any
): Dependency {
  if (isString(value)) value = { version: value };

  const {
    features = [],
    'default-features': defaultFeatures = true,
    optional = false,
    version,
    path,
    git,
    tag,
    branch = 'master',
    rev
  }: RawDependency = value;
  // console.log('value', value, version, path, git)

  if (!version && !path && !git) {
    throw new Error(
      `Invalid dependency "${name}", no version, path, or git specified. ${EXAMPLE}`
    );
  }

  let dependency;
  if (version) dependency = { version };
  else if (path) dependency = { path };
  else if (git) {
    if (rev) dependency = { git, rev };
    else if (tag) dependency = { git, tag };
    else dependency = { git, branch };
  }

  return { name, features, defaultFeatures, optional, ...dependency };
}
