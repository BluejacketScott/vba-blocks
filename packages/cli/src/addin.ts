import env from './env';
import { Project } from './project';
import { Target, Source, Reference } from './manifest';
import { run, unixJoin } from './utils';

export type Application = string;
export type Addin = string;

export const extensions: { [application: string]: string[] } = {
  excel: ['xlsx', 'xlsm', 'xlam']
};
export const addins: { [application: string]: string } = {
  excel: 'vba-blocks.xlsm' // TEMP vba-blocks.xlam
};

const byExtension: { [extension: string]: string } = {};
for (const [application, values] of Object.entries(extensions)) {
  for (const extension of values) {
    byExtension[extension] = application;
  }
}

/**
 * Import graph of src and references into given target
 */
export async function importGraph(
  project: Project,
  target: Target,
  graph: { src: Source[]; references: Reference[] }
): Promise<void> {
  const { application, addin, file } = getTargetInfo(project, target);
  const { src, references } = graph;

  await run(application, addin, 'Build.ImportGraph', {
    file,
    src,
    references
  });
}

/**
 * Export src and references from given target
 */
export async function exportTo(
  project: Project,
  target: Target,
  staging: string
): Promise<void> {
  const { application, addin, file } = getTargetInfo(project, target);

  const result = await run(application, addin, 'Build.ExportTo', {
    file,
    staging
  });
  console.log('export result', result);
}

/**
 * Get application, addin, and file for given target
 */
export function getTargetInfo(
  project: Project,
  target: Target
): { application: Application; addin: Addin; file: string } {
  const application = byExtension[target.type];
  if (!application) throw new Error(`Unsupported target type "${target.type}"`);

  const addin = unixJoin(env.addins, addins[application]);
  const file = unixJoin(project.paths.build, target.filename);

  return { application, addin, file };
}
