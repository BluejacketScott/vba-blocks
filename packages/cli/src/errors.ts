import chalk from 'chalk';
import env from './env';
import { cleanError, isString } from './utils';

import { Target } from './manifest';
import { Registration } from './sources';

export interface CliErrors {
  'unknown-command': (command: string) => string;

  'unsupported-source': (type: string) => string;
  'dependency-not-found': (dependency: string, registry: string) => string;
  'dependency-invalid-checksum': (registration: Registration) => string;

  'target-not-found': (target: Target) => string;
  'target-is-open': (target: Target, path: string) => string;
  'target-import-failed': (target: Target) => string;
  'target-restore-failed': (backup: string, file: string) => string;

  'resolve-failed': (details?: string) => string;

  'run-script-not-found': (path: string) => string;
}

export interface CliErrorOptions {
  code?: string;
  underlying?: Error;
}

export class CliError extends Error {
  code?: string;
  underlying?: Error;

  constructor(message: string, options: CliErrorOptions = {}) {
    if (options.underlying) {
      const underlying = cleanError(options.underlying);
      message += chalk`\n\n{dim ${underlying.message}}`;
    }

    super(message);

    this.code = options.code;
    this.underlying = options.underlying;
  }
}

// TODO Resolve typing issues so the following isn't so repetitive/verbose

export function unknownCommand(command: string): Error {
  const code = 'unknown-command';
  const message = env.reporter.errors[code](command);

  return new CliError(message, { code });
}

export function unsupportedSource(type: string): Error {
  const code = 'unsupported-source';
  const message = env.reporter.errors[code](type);

  return new CliError(message, { code });
}

export function dependencyNotFound(
  dependency: string,
  registry: string
): Error {
  const code = 'dependency-not-found';
  const message = env.reporter.errors[code](dependency, registry);

  return new CliError(message, { code });
}

export function dependencyInvalidChecksum(registration: Registration): Error {
  const code = 'dependency-invalid-checksum';
  const message = env.reporter.errors[code](registration);

  return new CliError(message, { code });
}

export function targetNotFound(target: Target): Error {
  const code = 'target-not-found';
  const message = env.reporter.errors[code](target);

  return new CliError(message, { code });
}

export function targetIsOpen(target: Target, path: string): Error {
  const code = 'target-is-open';
  const message = env.reporter.errors[code](target, path);

  return new CliError(message, { code });
}

export function targetImportFailed(target: Target, underlying: Error): Error {
  const code = 'target-import-failed';
  const message = env.reporter.errors[code](target);

  return new CliError(message, { code, underlying });
}

export function targetRestoreFailed(
  backup: string,
  file: string,
  underlying: Error
): Error {
  const code = 'target-restore-failed';
  const message = env.reporter.errors[code](backup, file);

  return new CliError(message, { code, underlying });
}

export function resolveFailed(details?: string): Error {
  const code = 'resolve-failed';
  let message = env.reporter.errors[code]();
  if (details) {
    message += `\n${details}`;
  }

  return new CliError(message, { code });
}

export function runScriptNotFound(path: string): Error {
  const code = 'run-script-not-found';
  const message = env.reporter.errors[code](path);

  return new CliError(message, { code });
}
