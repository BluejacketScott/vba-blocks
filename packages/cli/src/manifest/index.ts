import { join } from 'path';
import { ok } from 'assert';
import { parse as parseToml } from 'toml';
import { pathExists, readFile } from '../utils/fs';
import { Version } from './version';
import { Source, parseSrc } from './source';
import { Feature, parseFeatures } from './feature';
import { Dependency, parseDependencies } from './dependency';
import { Reference, parseReferences } from './reference';
import { Target, parseTargets } from './target';

export { Version, Source, Feature, Dependency, Reference, Target };

/**
 * @example
 * ```toml
 * [package]
 * name = "package-name"
 * version = "1.0.0-rc.1"
 * authors = ["Tim Hall <tim.hall.engr@gmail.com> (https://github.com/timhall)"]
 * 
 * [src]
 * A = "src/a.bas"
 * B = { path = "src/b.cls" }
 * C = { path = "src/c.frm", optional = true }
 * 
 * [features]
 * default = ["a"]
 * 
 * a = { src = ["C"] }
 * b = { dependencies = ["dictionary"] }
 * c = { references = ["Scripting"] }
 * 
 * [dependencies]
 * dictionary = "v1.4.1"
 * with-features = { version = "1.0.0", default-features = false, features = ["other"] }
 * from-path = { path = "packages/from-path" }
 * from-git-master = { git = "https://github.com/VBA-tools/VBA-Web.git" }
 * from-git-branch = { git = "https://github.com/VBA-tools/VBA-Web.git", branch = "beta" }
 * from-git-tag = { git = "https://github.com/VBA-tools/VBA-Web.git", tag = "v1.0.0" }
 * from-git-rev = { git = "https://github.com/VBA-tools/VBA-Web.git", rev = "a1b2c3d4" }
 * 
 * [references.Scripting]
 * version = "1.0"
 * guid = "{420B2830-E718-11CF-893D-00A0C9054228}"
 * optional = true
 * 
 * [[targets]]
 * type = "xlsm"
 * path = "targets/xlsm"
 * 
 * [[targets]]
 * name = "custom-name"
 * type = "xlam"
 * path = "targets/xlam"
 * ```
 */

export interface Snapshot {
  name: string;
  version: Version;
  dependencies: Dependency[];
}

export interface Metadata {
  authors: string[];
  publish: boolean;
  [name: string]: any;
}

export interface Manifest extends Snapshot {
  metadata: Metadata;
  src: Source[];
  references: Reference[];
  targets: Target[];
  features: Feature[];
  defaultFeatures: string[];
  dir: string;
}

const EXAMPLE = `Example vba-block.toml:

  [package]
  name = "my-package"
  version = "0.0.0"
  authors = ["..."]`;

export function parseManifest(value: any, dir: string): Manifest {
  ok(
    value && value.package,
    `[package] is a required field, with name, version, and authors specified. ${EXAMPLE}`
  );

  const { name, version, authors, publish = false } = value.package;

  ok(name, `[package] name is a required field. ${EXAMPLE}`);
  ok(version, `[package] version is a required field. ${EXAMPLE}`);
  ok(authors, `[package] authors is a required field. ${EXAMPLE}`);

  const src = parseSrc(value.src || {}, dir);
  const { features, defaultFeatures } = parseFeatures(value.features || {});
  const dependencies = parseDependencies(value.dependencies || {});
  const references = parseReferences(value.references || {});
  const targets = parseTargets(value.targets || [], name, dir);

  const metadata = { publish, ...value.package };

  return {
    name,
    version,
    metadata,
    src,
    features,
    defaultFeatures,
    dependencies,
    references,
    targets,
    dir
  };
}

export async function loadManifest(dir: string): Promise<Manifest> {
  const file = join(dir, 'vba-block.toml');
  if (!await pathExists(file)) {
    throw new Error(`vba-blocks.toml not found in "${dir}"`);
  }

  const raw = await readFile(file);
  const parsed = parseToml(raw.toString());
  const manifest = parseManifest(parsed, dir);

  return manifest;
}
