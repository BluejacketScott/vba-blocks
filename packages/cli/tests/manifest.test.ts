import { join, relative } from 'path';
import { unixPath } from '../src/utils';
import { Manifest, parseManifest, loadManifest } from '../src/manifest';
import { isPathDependency } from '../src/manifest/dependency';

const BASE_MANIFEST = {
  package: { name: 'package-name', version: '1.0.0', authors: ['Tim Hall'] }
};

const FIXTURES = join(__dirname, 'fixtures');

test('loads valid package metadata', () => {
  expect(normalize(parseManifest(BASE_MANIFEST, FIXTURES))).toMatchSnapshot();
});

test('throws for invalid package metadata', () => {
  expect(() => parseManifest({}, FIXTURES)).toThrow();
  expect(() => parseManifest({ package: {} }, FIXTURES)).toThrow();
  expect(() =>
    parseManifest({ package: { name: 'package-name' } }, FIXTURES)
  ).toThrow();
  expect(() =>
    parseManifest({ package: { version: '1.0.0' } }, FIXTURES)
  ).toThrow();
  expect(() =>
    parseManifest({ package: { authors: ['Tim Hall'] } }, FIXTURES)
  ).toThrow();
});

test('loads valid sources', () => {
  const value = {
    ...BASE_MANIFEST,
    src: {
      A: 'src/a.bas',
      B: { path: 'src/b.cls' },
      C: { path: 'src/c.frm', optional: true }
    }
  };

  expect(normalize(parseManifest(value, FIXTURES))).toMatchSnapshot();
});

test('throws for invalid sources', () => {
  const value = {
    ...BASE_MANIFEST,
    src: {
      missing_path: { optional: true }
    }
  };

  expect(() => parseManifest(value, FIXTURES)).toThrow();
});

test('loads valid features', () => {
  const value = {
    ...BASE_MANIFEST,
    features: {
      default: ['a', 'b'],
      a: { src: ['A'] },
      b: { dependencies: ['B'] },
      c: { references: ['C'] }
    }
  };

  expect(normalize(parseManifest(value, FIXTURES))).toMatchSnapshot();
});

test('throws for invalid features', () => {
  let value: any = { ...BASE_MANIFEST, features: { a: { src: 'A' } } };
  expect(() => parseManifest(value, FIXTURES)).toThrow();

  value = { ...BASE_MANIFEST, features: { b: { dependencies: 'B' } } };
  expect(() => parseManifest(value, FIXTURES)).toThrow();

  value = { ...BASE_MANIFEST, features: { c: { references: 'C' } } };
  expect(() => parseManifest(value, FIXTURES)).toThrow();
});

test('loads valid dependencies', () => {
  const value = {
    ...BASE_MANIFEST,
    dependencies: {
      a: '^1.0.0',
      b: {
        version: '^2.0.0',
        optional: true,
        features: ['a', 'b'],
        'default-features': false
      },
      c: { path: 'packages/c' },
      d: { git: 'https://github.com/VBA-tools/VBA-Web' },
      e: { git: 'https://github.com/VBA-tools/VBA-Web', branch: 'next' },
      f: { git: 'https://github.com/VBA-tools/VBA-Web', tag: 'v1.0.0' },
      g: { git: 'https://github.com/VBA-tools/VBA-Web', rev: 'a1b2c3d4' }
    }
  };

  expect(normalize(parseManifest(value, FIXTURES))).toMatchSnapshot();
});

test('throws for invalid dependencies', () => {
  const value: any = { ...BASE_MANIFEST, dependencies: { a: {} } };
  expect(() => parseManifest(value, FIXTURES)).toThrow();
});

test('load valid references', () => {
  const value = {
    ...BASE_MANIFEST,
    references: {
      a: {
        version: '1.0',
        guid: '{420B2830-E718-11CF-893D-00A0C9054228}',
        optional: true
      }
    }
  };

  expect(normalize(parseManifest(value, FIXTURES))).toMatchSnapshot();
});

test('throws for invalid references', () => {
  let value: any = { ...BASE_MANIFEST, references: { a: {} } };
  expect(() => parseManifest(value, FIXTURES)).toThrow();

  value = { ...BASE_MANIFEST, references: { b: { version: '1.0.0' } } };
  expect(() => parseManifest(value, FIXTURES)).toThrow();

  value = { ...BASE_MANIFEST, references: { c: { version: '1.0' } } };
  expect(() => parseManifest(value, FIXTURES)).toThrow();

  value = {
    ...BASE_MANIFEST,
    references: {
      d: { version: '1.0', guid: 'aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa' }
    }
  };
  expect(() => parseManifest(value, FIXTURES)).toThrow();
});

test('loads valid targets', () => {
  const value = {
    ...BASE_MANIFEST,
    targets: [
      { type: 'xlsm', path: 'targets/xlsm' },
      { name: 'addin', type: 'xlam', path: 'targets/xlam' }
    ]
  };

  expect(normalize(parseManifest(value, FIXTURES))).toMatchSnapshot();
});

test('throws for invalid targets', () => {
  let value: any = { ...BASE_MANIFEST, targets: {} };
  expect(() => parseManifest(value, FIXTURES)).toThrow();

  value = { ...BASE_MANIFEST, targets: [{}] };
  expect(() => parseManifest(value, FIXTURES)).toThrow();

  value = { ...BASE_MANIFEST, targets: [{ type: 'xlsm' }] };
  expect(() => parseManifest(value, FIXTURES)).toThrow();
});

test('loads and parses manifest', async () => {
  const manifest = await loadManifest(join(FIXTURES, 'project'));
  expect(normalize(manifest, join(FIXTURES, 'project'))).toMatchSnapshot();
});

function normalize(
  manifest: Manifest,
  relativeTo: string = FIXTURES
): Manifest {
  manifest.dir = normalizePath(manifest.dir, relativeTo);

  for (const src of manifest.src) {
    src.path = normalizePath(src.path, relativeTo);
  }
  for (const target of manifest.targets) {
    target.path = normalizePath(target.path, relativeTo);
  }
  for (const dependency of manifest.dependencies) {
    if (isPathDependency(dependency)) {
      dependency.path = normalizePath(dependency.path, relativeTo);
    }
  }

  return manifest;
}

function normalizePath(path: string, relativeTo: string = FIXTURES): string {
  return unixPath(relative(relativeTo, path));
}
