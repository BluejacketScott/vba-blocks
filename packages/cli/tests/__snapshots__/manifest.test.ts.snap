// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`load valid references 1`] = `
Object {
  "defaultFeatures": Array [],
  "dependencies": Array [],
  "dir": ".",
  "features": Array [],
  "metadata": Object {
    "authors": Array [
      "Tim Hall",
    ],
    "name": "package-name",
    "publish": false,
    "version": "1.0.0",
  },
  "name": "package-name",
  "references": Array [
    Object {
      "guid": "{420B2830-E718-11CF-893D-00A0C9054228}",
      "major": 1,
      "minor": 0,
      "name": "a",
      "optional": true,
      "version": "1.0",
    },
  ],
  "src": Array [],
  "targets": Array [],
  "version": "1.0.0",
}
`;

exports[`loads and parses manifest 1`] = `
Object {
  "defaultFeatures": Array [
    "a",
    "b",
  ],
  "dependencies": Array [
    Object {
      "defaultFeatures": true,
      "features": Array [],
      "name": "dictionary",
      "optional": true,
      "registry": "vba-blocks",
      "version": "1.4.1",
    },
    Object {
      "defaultFeatures": false,
      "features": Array [
        "dictionary-scripting",
      ],
      "git": "https://github.com/VBA-tools/VBA-Web",
      "name": "web",
      "optional": false,
      "tag": "v4.1.3",
    },
  ],
  "dir": ".",
  "features": Array [
    Object {
      "dependencies": Array [],
      "name": "a",
      "references": Array [],
      "src": Array [
        "b",
      ],
    },
    Object {
      "dependencies": Array [
        "dictionary",
      ],
      "name": "b",
      "references": Array [],
      "src": Array [],
    },
    Object {
      "dependencies": Array [],
      "name": "c",
      "references": Array [
        "scripting",
      ],
      "src": Array [],
    },
  ],
  "metadata": Object {
    "authors": Array [
      "Tim Hall <tim.hall.engr@gmail.com>",
    ],
    "name": "testing",
    "publish": true,
    "version": "0.1.0",
  },
  "name": "testing",
  "references": Array [
    Object {
      "guid": "{420B2830-E718-11CF-893D-00A0C9054228}",
      "major": 1,
      "minor": 0,
      "name": "Scripting",
      "optional": true,
      "version": "1.0",
    },
  ],
  "src": Array [
    Object {
      "name": "a",
      "optional": false,
      "path": "src/a.bas",
    },
    Object {
      "name": "b",
      "optional": true,
      "path": "src/b.cls",
    },
  ],
  "targets": Array [
    Object {
      "name": "testing",
      "path": "targets/xlsm",
      "type": "xlsm",
    },
  ],
  "version": "0.1.0",
}
`;

exports[`loads valid dependencies 1`] = `
Object {
  "defaultFeatures": Array [],
  "dependencies": Array [
    Object {
      "defaultFeatures": true,
      "features": Array [],
      "name": "a",
      "optional": false,
      "registry": "vba-blocks",
      "version": "^1.0.0",
    },
    Object {
      "defaultFeatures": false,
      "features": Array [
        "a",
        "b",
      ],
      "name": "b",
      "optional": true,
      "registry": "vba-blocks",
      "version": "^2.0.0",
    },
    Object {
      "defaultFeatures": true,
      "features": Array [],
      "name": "c",
      "optional": false,
      "path": "packages/c",
    },
    Object {
      "branch": "master",
      "defaultFeatures": true,
      "features": Array [],
      "git": "https://github.com/VBA-tools/VBA-Web",
      "name": "d",
      "optional": false,
    },
    Object {
      "branch": "next",
      "defaultFeatures": true,
      "features": Array [],
      "git": "https://github.com/VBA-tools/VBA-Web",
      "name": "e",
      "optional": false,
    },
    Object {
      "defaultFeatures": true,
      "features": Array [],
      "git": "https://github.com/VBA-tools/VBA-Web",
      "name": "f",
      "optional": false,
      "tag": "v1.0.0",
    },
    Object {
      "defaultFeatures": true,
      "features": Array [],
      "git": "https://github.com/VBA-tools/VBA-Web",
      "name": "g",
      "optional": false,
      "rev": "a1b2c3d4",
    },
  ],
  "dir": ".",
  "features": Array [],
  "metadata": Object {
    "authors": Array [
      "Tim Hall",
    ],
    "name": "package-name",
    "publish": false,
    "version": "1.0.0",
  },
  "name": "package-name",
  "references": Array [],
  "src": Array [],
  "targets": Array [],
  "version": "1.0.0",
}
`;

exports[`loads valid features 1`] = `
Object {
  "defaultFeatures": Array [
    "a",
    "b",
  ],
  "dependencies": Array [],
  "dir": ".",
  "features": Array [
    Object {
      "dependencies": Array [],
      "name": "a",
      "references": Array [],
      "src": Array [
        "A",
      ],
    },
    Object {
      "dependencies": Array [
        "B",
      ],
      "name": "b",
      "references": Array [],
      "src": Array [],
    },
    Object {
      "dependencies": Array [],
      "name": "c",
      "references": Array [
        "C",
      ],
      "src": Array [],
    },
  ],
  "metadata": Object {
    "authors": Array [
      "Tim Hall",
    ],
    "name": "package-name",
    "publish": false,
    "version": "1.0.0",
  },
  "name": "package-name",
  "references": Array [],
  "src": Array [],
  "targets": Array [],
  "version": "1.0.0",
}
`;

exports[`loads valid package metadata 1`] = `
Object {
  "defaultFeatures": Array [],
  "dependencies": Array [],
  "dir": ".",
  "features": Array [],
  "metadata": Object {
    "authors": Array [
      "Tim Hall",
    ],
    "name": "package-name",
    "publish": false,
    "version": "1.0.0",
  },
  "name": "package-name",
  "references": Array [],
  "src": Array [],
  "targets": Array [],
  "version": "1.0.0",
}
`;

exports[`loads valid sources 1`] = `
Object {
  "defaultFeatures": Array [],
  "dependencies": Array [],
  "dir": ".",
  "features": Array [],
  "metadata": Object {
    "authors": Array [
      "Tim Hall",
    ],
    "name": "package-name",
    "publish": false,
    "version": "1.0.0",
  },
  "name": "package-name",
  "references": Array [],
  "src": Array [
    Object {
      "name": "A",
      "optional": false,
      "path": "src/a.bas",
    },
    Object {
      "name": "B",
      "optional": false,
      "path": "src/b.cls",
    },
    Object {
      "name": "C",
      "optional": true,
      "path": "src/c.frm",
    },
  ],
  "targets": Array [],
  "version": "1.0.0",
}
`;

exports[`loads valid targets 1`] = `
Object {
  "defaultFeatures": Array [],
  "dependencies": Array [],
  "dir": ".",
  "features": Array [],
  "metadata": Object {
    "authors": Array [
      "Tim Hall",
    ],
    "name": "package-name",
    "publish": false,
    "version": "1.0.0",
  },
  "name": "package-name",
  "references": Array [],
  "src": Array [],
  "targets": Array [
    Object {
      "name": "package-name",
      "path": "targets/xlsm",
      "type": "xlsm",
    },
    Object {
      "name": "addin",
      "path": "targets/xlam",
      "type": "xlam",
    },
  ],
  "version": "1.0.0",
}
`;
